### ISSUE TITLE
ObjectMapper canSerialize function has performance loss after 2.4.4

### ISSUE THREAD

**Issue by jjmatrix** [FasterXML/jackson-databind#1180]
Last week，we change version to 2.6.4，but we found there are performance loss in stress test。

Finnally, i found it's in canSerialize function. 

```java
 protected JsonSerializer<Object> _findExplicitUntypedSerializer(Class<?> runtimeType)
        throws JsonMappingException
    {        
        // Fast lookup from local lookup thingy works?
        JsonSerializer<Object> ser = _knownSerializers.untypedValueSerializer(runtimeType);
        if (ser == null) {
            // If not, maybe shared map already has it?
            ser = _serializerCache.untypedValueSerializer(runtimeType);
            if (ser == null) {
                ser = _createAndCacheUntypedSerializer(runtimeType);
            }
        }
        /* 18-Sep-2014, tatu: This is unfortunate patch over related change
         *    that pushes creation of "unknown type" serializer deeper down
         *    in BeanSerializerFactory; as a result, we need to "undo" creation
         *    here.
         */
        if (isUnknownTypeSerializer(ser)) {
            return null;
        }
        return ser;
    }
```

Here, the key untypedValueSerializer use to look for JsonSerializer and _createAndCacheUntypedSerializer use is different. so it create a new JsonSerializer every time, even worse，it is synchronized.

JavaType type = _config.constructType(rawType);

I don't know why do this conversion?

**Comment by cowtowncoder** [comment]
First of all, I would recommend removing use of `canSerialize()` altogether, if possible.
It is questionable whether it provides much value as things are.

But second, why do you think this is the root cause? New serializer should not be constructed, unless there are some other issue with code, like not reusing `ObjectMapper` -- serializers. But even if a new instance was constructed once, further calls to access and use such serializer should involve caching. So it is not clear how this could result in observable performance issue.

One thing that could help (although not very likely) is to use latest 2.6 patch, 2.6.5. At least worth trying.

**Comment by jjmatrix** [comment]
I can't remove the use of `canSerialize()`, because Spring framework 4.2.4 `AbstractJackson2HttpMessageConverter` use `canSerialize()`.

In addition, the problem is that it can't hit the cache when further calls to access and use such serializer, I write a test case before.
```
@Test
public void testCanSerialize(){
    TypeKey typeKeyByCls = new TypeKey(ResponseVO.class, false);
 
    SerializationConfig config = objectMapper.getSerializationConfig();
    JavaType javaType = config.constructType(ResponseVO.class);
    TypeKey typeKeyByType = new TypeKey(javaType, false);
 
    assertTrue(!typeKeyByCls.equals(typeKeyByType));
}
``` 

I found the source code of 2.6.5 is the same in `canSerialize()`.

I wonder if this is a bug?

**Comment by cowtowncoder** [comment]
@jjmatrix Ok (wrt Spring using it -- JAX-RS implementations do the same, unfortunately).

Regarding the test: it is not showing an exact bug since two representations are not meant to be comparable directly (that can be costly operation and `TypeKey` explicitly avoids trying to unify handling; cost comes from generic type, structure checks), but it does show how caching might fail to "stick", causing issue you see.

But I think that something was fixed in the way lookups worked for 2.7, although was unable to see a filed issue -- code in `_createAndCacheUntypedSerializer` of `SerializerProvider` (line 1175) notes:

```
        if (ser != null) {
            // 21-Dec-2015, tatu: Best to cache for both raw and full-type key
            _serializerCache.addAndResolveNonTypedSerializer(rawType, fullType, ser, this);
        }
```

comment which is missing from 2.6 sources. Next method down does have a question regarding applicabilty, so it could be incomplete fix.
Now: since we just released 2.6.6, which is planned to be the last full patch release for 2.6 branch, I think fix would have to be in 2.7 in any case.
